doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/NamesForIndexEntriesSinkWrapper.java

```diff
@@ -0,0 +1,136 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.doxia.sink.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.maven.doxia.index.IndexEntry;
+import org.apache.maven.doxia.index.SectionLocation;
+import org.apache.maven.doxia.macro.toc.TocMacro;
+import org.apache.maven.doxia.parser.SinkWrapper;
+import org.apache.maven.doxia.sink.Sink;
+import org.apache.maven.doxia.sink.SinkEventAttributes;
+
+/**
+ * Sink wrapper which emits anchors for each section represented by the given {@link IndexEntry}.
+ * @see TocMacro
+ */
+public class NamesForIndexEntriesSinkWrapper extends SinkWrapper {
+
+    private final Map<SectionLocation, IndexEntry> entryByLocation;
+    private SectionLocation currentLocation;
+
+    public NamesForIndexEntriesSinkWrapper(Sink delegate, IndexEntry index) {
+        super(delegate);
+        entryByLocation = new HashMap<>();
+        populateMap(entryByLocation, index);
+        this.currentLocation = new SectionLocation();
+    }
+
+    private static void populateMap(Map<SectionLocation, IndexEntry> entryByLocation, IndexEntry entry) {
+        entryByLocation.put(entry.getLocation(), entry);
+        entry.getChildEntries().forEach(e -> populateMap(entryByLocation, e));
+    }
+
+    @Override
+    public void section(int level, SinkEventAttributes attributes) {
+        onStartSection(level);
+        super.section(level, attributes);
+    }
+
+    @Override
+    public void section_(int level) {
+        super.section_(level);
+        onEndSection();
+    }
+
+    @Override
+    public void section1() {
+        super.section1();
+        onStartSection(1);
+    }
+
+    @Override
+    public void section1_() {
+        super.section1_();
+        onEndSection();
+    }
+
+    @Override
+    public void section2() {
+        super.section2();
+        onStartSection(2);
+    }
+
+    @Override
+    public void section2_() {
+        super.section2_();
+        onEndSection();
+    }
+
+    @Override
+    public void section3() {
+        super.section3();
+        onStartSection(3);
+    }
+
+    @Override
+    public void section3_() {
+        super.section3_();
+        onEndSection();
+    }
+
+    @Override
+    public void section4() {
+        super.section4();
+        onStartSection(4);
+    }
+
+    @Override
+    public void section4_() {
+        super.section4_();
+        onEndSection();
+    }
+
+    @Override
+    public void section5() {
+        super.section5();
+        onStartSection(5);
+    }
+
+    @Override
+    public void section5_() {
+        super.section5_();
+        onEndSection();
+    }
+
+    private void onStartSection(int level) {
+        currentLocation = currentLocation.createChildSection(level);
+        IndexEntry entry = entryByLocation.get(currentLocation);
+        if (entry != null) {
```

Comment:
Probably a guard similar to https://github.com/apache/maven-doxia-sitetools/pull/58/files#diff-dabe16ceeba5dc7cb0a15e30dcb7c204a78f4763346f016ab7639ab2f59e78d4L250 needs to be introduced to at least prevent double anchors for APT format.